### PR TITLE
Refactor the recursive solver a bit

### PR DESCRIPTION
**Pull request by jackh726**
This is about 95% of the work needed to put the recursive solver in a separate crate.

Everything in `chalk_solve::recursive::*` can get moved. `recursive.rs` is mostly the equivalent of the `slg` submodule.

### CODE REVIEW

File: chalk-engine/src/forest.rs [hunk 1]
```diff
@@ -54,89 +53,6 @@ impl<I: Interner, C: Context<I>> Forest<I, C> {
             _context: std::marker::PhantomData::<C>,
         }
     }
-
-    /// Solves a given goal, producing the solution. This will do only
-    /// as much work towards `goal` as it has to (and that works is
-    /// cached for future attempts).
-    pub fn solve(
```
Comment (jackh726) [inline chalk-engine/src/forest.rs:61]: As sort of a drive-by change, I moved these two functions `solve` and `solve_multiple` out of `Forest` (and `chalk-engine`) and into `chalk_solve::solve`. For `solve`, it honestly just makes more sense, since this is the *only* place solution is needed. For `solve_multiple`, there's a little work done here, but really I think these are implementation details akin to `aggregate`.

File: chalk-solve/src/recursive.rs [hunk 2]
```diff
@@ -26,6 +33,104 @@ pub(crate) struct RecursiveContext<I: Interner> {
     caching_enabled: bool,
 }
 
+pub(crate) struct RecursiveInferenceTableImpl<I: Interner> {
+    pub(crate) infer: InferenceTable<I>,
+}
+
+impl<I: Interner> RecursiveInferenceTable<I> for RecursiveInferenceTableImpl<I> {
```
Comment (jackh726) [inline chalk-solve/src/recursive.rs:40]: I started to wonder if maybe it's worth having this notion of an `InferenceTable` (as a trait) in `chalk_ir`. A lot of these functions are also used for slg. Notable differences include `invert_then_canonicalize` and actually using the `free_vars` from `canonicalize`

File: chalk-solve/src/recursive/lib.rs [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    }
+
+    pub fn update_from(&mut self, minimums: Minimums) {
+        self.positive = ::std::cmp::min(self.positive, minimums.positive);
+    }
+}
+
+/// A (possible) solution for a proposed goal.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Solution<I: Interner> {
```
Comment (jackh726) [inline chalk-solve/src/recursive/lib.rs:30]: This is mostly just copied from `chalk_solve::solve`. But I didn't really think of a good way to handle this. We can make `SearchGraph` generic over `Solution`, but `Fulfill:solve` needs to know about these. I wish it was easier to split out `solve` from `Fulfill` (similar to how we split out `aggregate` from `chalk-engine`, but the `solve` function needs to know a *lot* about the internals of `Fulfill` that we would have to expose publicly.

File: chalk-solve/src/solve.rs [hunk 4]
```diff
@@ -57,100 +60,6 @@ impl<I: Interner> Solution<I> {
         }
     }
 
-    /// There are multiple candidate solutions, which may or may not agree on
```
Comment (jackh726) [inline chalk-solve/src/solve.rs:60]: On the other hand (continued from where these are copied to), the fact that several of these functions are only used in the recursive solver makes me wonder if this is mostly okay anyways.

File: chalk-solve/src/solve.rs [hunk 5]
```diff
@@ -316,10 +225,32 @@ impl<I: Interner> Solver<I> {
                 expected_answers,
             } => {
                 let ops = SlgContextOps::new(program, *max_size, *expected_answers);
-                forest.solve(&ops, goal, || true)
+                ops.make_solution(goal, forest.iter_answers(&ops, goal), || true)
             }
             #[cfg(feature = "recursive-solver")]
-            SolverImpl::Recursive(ctx) => ctx.solver(program).solve_root_goal(goal).ok(),
+            SolverImpl::Recursive(ctx) => {
+                ctx.solver(program)
```
Comment (jackh726) [inline chalk-solve/src/solve.rs:232]: *But* (continued from above still), this is a bit annoying.

Comment (zaharidichev) [reply chalk-solve/src/solve.rs:232]: Yes.. that is indeed a bit annoying. Nevertheless if what you suggested happens (chalk engine + chalk_recursive depending on chalk solve), this one solver struct can go away correct?

File: chalk-solve/src/recursive.rs (hunk 2 above)
Comment (nikomatsakis) [reply chalk-solve/src/recursive.rs:40]: I think the inference code *itself* should maybe move to chalk-ir, actually, but we should discuss, especially in light of your idea to "invert" the dependency layers.

File: chalk-solve/src/recursive/lib.rs (hunk 3 above)
Comment (nikomatsakis) [reply chalk-solve/src/recursive/lib.rs:30]: I feel.. hmm.. a bit confused about this. I guess I think I would expect this type to potentially be in chalk-ir -- is it really *specific* to the recursive solver? Don't we use a similar type for SLG? (I guess as the *output* from aggregation?) 

I think my take is that aggregation probably belongs in with the SLG solver in the end, although it is definitely a distinct "layer".
